vm/src/protocol/number.rs

```diff
@@ -441,71 +459,101 @@ impl PyNumber<'_> {
         self.class().slots.as_number.int.load().map(|f| {
             let ret = f(self, vm)?;
-            let value = if !ret.class().is(PyInt::class(&vm.ctx)) {
+
+            if let Some(ret) = ret.downcast_ref_if_exact::<PyInt>(vm) {
+                return Ok(ret.to_owned());
+            }
+
+            let ret_class = ret.class().to_owned();
+            if let Some(ret) = ret.downcast_ref::<PyInt>() {
                 warnings::warn(
                     vm.ctx.exceptions.deprecation_warning,
                     format!(
                         "__int__ returned non-int (type {}).  \
-                The ability to return an instance of a strict subclass of int \
-                is deprecated, and may be removed in a future version of Python.",
-                        ret.class()
+                    The ability to return an instance of a strict subclass of int \
+                    is deprecated, and may be removed in a future version of Python.",
+                        ret_class
                     ),
                     1,
                     vm,
                 )?;
-                vm.ctx.new_bigint(ret.as_bigint())
+
+                Ok(ret.to_owned())
             } else {
-                ret
-            };
-            Ok(value)
+                Err(vm.new_type_error(format!(
+                    "{}.__int__ returned non-int(type {})",
+                    self.class(),
+                    ret_class
+                )))
+            }
         })
     }
 
     #[inline]
     pub fn index(self, vm: &VirtualMachine) -> Option<PyResult<PyIntRef>> {
         self.class().slots.as_number.index.load().map(|f| {
             let ret = f(self, vm)?;
-            let value = if !ret.class().is(PyInt::class(&vm.ctx)) {
+
+            if let Some(ret) = ret.downcast_ref_if_exact::<PyInt>(vm) {
+                return Ok(ret.to_owned());
+            }
+
+            let ret_class = ret.class().to_owned();
+            if let Some(ret) = ret.downcast_ref::<PyInt>() {
                 warnings::warn(
                     vm.ctx.exceptions.deprecation_warning,
                     format!(
                         "__index__ returned non-int (type {}).  \
-                The ability to return an instance of a strict subclass of int \
-                is deprecated, and may be removed in a future version of Python.",
-                        ret.class()
+                    The ability to return an instance of a strict subclass of int \
+                    is deprecated, and may be removed in a future version of Python.",
+                        ret_class
                     ),
                     1,
                     vm,
                 )?;
-                vm.ctx.new_bigint(ret.as_bigint())
+
+                Ok(ret.to_owned())
             } else {
-                ret
-            };
-            Ok(value)
+                Err(vm.new_type_error(format!(
+                    "{}.__index__ returned non-int(type {})",
```

Comment:
Is this better to be put in here? I thought only python implementations can make this mis-implementation, so it will be a wrapper function.

Reply:
Do you mean that we do not need to exception handling for that situation or move them to something where?

Reply:
Remind) @youknowone  What more should I do for this PR to be merged?